csi.proto

```diff
@@ -99,6 +93,58 @@ message GetPluginInfoResponse {
 }
 ////////
 ////////
+message GetPluginCapabilitiesRequest {
+  // The API version assumed by the CO. This is a REQUIRED field.
+  Version version = 1;
+}
+
+message GetPluginCapabilitiesResponse {
+  // All the capabilities that the controller service supports. This
+  // field is OPTIONAL.
+  // If any of the `CONTROLLER_xxx` RPC capabilities are specified then
+  // it is REQUIRED for a plugin to also report a `CONTROLLER_SERVICE`
+  // service capability as well.
+  repeated PluginCapability capabilities = 2;
+}
+
+// Specifies a capability of the plugin.
+message PluginCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+      CONTROLLER_CREATE_DELETE_VOLUME = 1;
+      CONTROLLER_PUBLISH_UNPUBLISH_VOLUME = 2;
+      CONTROLLER_LIST_VOLUMES = 3;
+      CONTROLLER_GET_CAPACITY = 4;
+    }
+
+    Type type = 1;
+  }
+
+  message Service {
+    enum Type {
+      UNKNOWN = 0;
+      CONTROLLER_SERVICE = 1;
+    }
+  }
+
+  oneof type {
+    // RPC that the plugin supports.
+    RPC rpc = 1;
+    // Service that the plugin supports.
+    Service service = 2;
```

Comment:
It looks like you're trying to solve the problem of having a way to enable/disable the entire `CONTROLLER` service and have a way to control individual features of the `CONTROLLER` service.

But I'm not quite sure how this will work. If I want to enable the controller service but only have some of the controller capabilities, how would I do that?

I do like having a single PluginCapability instead of each service having it's own. But I'm not sure of a cleaner way to solve the problem above.

Reply:
But I'm not quite sure how this will work. If I want to enable the controller service but only have some of the controller capabilities, how would I do that?

see https://github.com/container-storage-interface/spec/pull/184/files#diff-958e7270f96f5407d7d980f500805b1bR360